core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -196,34 +199,39 @@ protected boolean resolveIndex() {
 
         try {
             // add the response
-            final WriteResult<DeleteResponse> writeResult = TransportDeleteAction.executeDeleteRequestOnPrimary(deleteRequest, indexShard);
-            DeleteResponse deleteResponse = writeResult.getResponse();
-            location = locationToSync(location, writeResult.getLocation());
-            setResponse(item, new BulkItemResponse(item.id(), OP_TYPE_DELETE, deleteResponse));
+            WriteResult<DeleteResponse> writeResult = TransportDeleteAction.executeDeleteRequestOnPrimary(deleteRequest, indexShard);
+            if (writeResult.success()) {
+                DeleteResponse deleteResponse = writeResult.getResponse();
+                location = locationToSync(location, writeResult.getLocation());
+                setResponse(item, new BulkItemResponse(item.id(), OP_TYPE_DELETE, deleteResponse));
+            } else {
+                logFailure(writeResult.getFailure(), "delete", request.shardId(), deleteRequest);
+                // if its a conflict failure, and we already executed the request on a primary (and we execute it
+                // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
+                // then just use the response we got from the successful execution
+                if (item.getPrimaryResponse() != null && isConflictException(writeResult.getFailure())) {
+                    setResponse(item, item.getPrimaryResponse());
+                } else {
+                    setResponse(item, new BulkItemResponse(item.id(), OP_TYPE_DELETE,
+                        new BulkItemResponse.Failure(request.index(), deleteRequest.type(), deleteRequest.id(), writeResult.getFailure())));
+                }
+            }
         } catch (Exception e) {
-            // rethrow the failure if we are going to retry on primary and let parent failure to handle it
             if (retryPrimaryException(e)) {
                 // restore updated versions...
                 for (int j = 0; j < requestIndex; j++) {
                     applyVersion(request.items()[j], preVersions[j], preVersionTypes[j]);
                 }
-                throw (ElasticsearchException) e;
+                throw e;
```

Comment:
this doesn't force to add a throws Exception because `TransportDeleteAction#executeDeleteRequestOnPrimary` method didn't explicitly specify it throws Exception in the method signature.